### PR TITLE
Add EIP tests, and move TagMap to morphodict

### PR DESCRIPTION
**Pull request by andrewdotn**

### CODE REVIEW

File: src/CreeDictionary/API/search/eip_crk.py [hunk 1]
```diff
@@ -20,6 +21,110 @@ def get_noun_tags(inflectional_category):
...
+    ("+Imm", ("+Imp", "+Imm"), 1),  # Immediate imperative
+    ("+Del", ("+Imp", "+Del"), 1),  # Delayed imperative
+    ("+Fut", ("PV/wi+", "+Ind"), 1),  # Future
+    # ("+Fut", "PV/wi+", 1)  # Also accept PV/wi without indicative as future
+    # Note that these crk features as disjoint, but both are needed for the eng feature
+    ("+Def", ("PV/ka+", "+Ind"), 1),
+    ("+Inf", ("PV/ka+", "+Cnj"), 1),
+    (TagMap.DEFAULT, "+Ind", 1),
+    # "+Inf": ("PV/ta+", "+Cnj")  # future definite
+    *passthrough_tags_to_tuples(verb_passthrough_tags)
```
Comment (eddieantonio) [inline src/CreeDictionary/API/search/eip_crk.py:120]: Oh, I like this pattern!

File: src/CreeDictionary/API/search/test_eip_crk.py [hunk 2]
```diff
@@ -1,24 +1,72 @@
...
+                "lemma": "wâpamêw",
+                "expected_inflection": "nikî-wâpamikwak",
+            },
+        ],
+        [
+            "my little bears",
+            {
+                "expected_query_terms": ["bear"],
+                "expected_new_tags": ["+N", "+Der/Dim", "+Px1Sg", "+Pl"],
+                "lemma": "maskwa",
```
Comment (eddieantonio) [inline src/CreeDictionary/API/search/test_eip_crk.py:35]: I'm curious: should the system make the lemma [maskosis](https://itwewina.altlab.app/word/maskosis/?paradigm-size=FULL). Its inflection is different: https://itwewina.altlab.app/search?q=nimaskosisak

File: src/CreeDictionary/API/search/test_eip_crk.py [hunk 3]
```diff
@@ -1,24 +1,72 @@
...
+            target_language_keyword_match=params["expected_query_terms"],
+        )
     )
 
     eip_search.inflect_search_results()
 
-    # kî-pimitâcimowak = “they crawled”
-    assert list(search_run.unsorted_results())[0].wordform.text == "kî-pimitâcimowak"
+    assert (
+        list(search_run.unsorted_results())[0].wordform.text
```
Comment (eddieantonio) [inline src/CreeDictionary/API/search/test_eip_crk.py:70]: It will always be the first, unsorted result?

File: src/morphodict/analysis/tag_map.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
+class UnknownTagError(KeyError):
+    """Raised when TagMap encounters an unknown tag"""
+
+
+class TagMap:
+    """Map between different sets of tags used by related FSTs
```
Comment (eddieantonio) [inline src/morphodict/analysis/tag_map.py:6]: 👍🏼  for docstrings!

File: src/morphodict/analysis/tag_map.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        self._multi_mappings = []
+        self._tag_mapping = {}
+        self._precedences = {}
+        self._defaults_by_precedence = {}
+
+        for input_tag_spec, output_tag_spec, prec in tag_definitions:
+            if isinstance(input_tag_spec, tuple):
+                if output_tag_spec == TagMap.COPY_TAG_NAME:
+                    raise Exception(
+                        f"Error: cannot use copy with multi-tags: {input_tag_spec}"
```
Comment (eddieantonio) [inline src/morphodict/analysis/tag_map.py:70]: Is there a better exception class for this? Or can a more specific exception be made for this?

File: src/morphodict/analysis/tag_map.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                    )
+                self._multi_mappings.append((input_tag_spec, output_tag_spec, prec))
+            elif input_tag_spec == TagMap.DEFAULT:
+                if prec in self._defaults_by_precedence:
+                    raise Exception(
+                        f"Error: multiple defaults supplied for precedence {prec}: {self._defaults_by_precedence[prec]}, {output_tag_spec}"
+                    )
```
Comment (eddieantonio) [inline src/morphodict/analysis/tag_map.py:77]: Ditto here. Something like `InvalidTagMapError` maybe?

File: src/morphodict/analysis/tag_map.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                    assert input_tag_spec.startswith(
+                        "+"
+                    ), f"expected tag to start with + but did not: {input_tag_spec}"
+                    output_tag_spec = input_tag_spec[1:] + "+"
+                self._tag_mapping[input_tag_spec] = output_tag_spec
+
+            if output_tag_spec is not None:
+                if output_tag_spec in self._precedences:
+                    if prec != self._precedences[output_tag_spec]:
+                        raise Exception(
```
Comment (eddieantonio) [inline src/morphodict/analysis/tag_map.py:90]: as above

**Review by eddieantonio — APPROVED** [review]
LGTM. I have a few questions, but nothing blocking!

File: src/CreeDictionary/API/search/test_eip_crk.py (hunk 3 above)
Comment (andrewdotn) [reply src/CreeDictionary/API/search/test_eip_crk.py:70]: Yeah, that’s what we’re testing. We create a search run, add a single result to it, tell the EIP search to do its inflection magic, and make sure that the single result has then been updated.

**Pull request merged**